core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
-		lastHash:   lastHash,
-		lastHeader: lastHeader,
-	}, nil
-}
-
-func (hc *HeaderChain) ValidateHeaderChain(chain []*types.Header, checkFreq int) (int, error) {
-	// Do a sanity check that the provided chain is actually ordered and linked
-	for i := 1; i < len(chain); i++ {
-		if chain[i].Number[types.QuaiNetworkContext].Uint64() != chain[i-1].Number[types.QuaiNetworkContext].Uint64()+1 {
-			hash := chain[i].Hash()
-			parentHash := chain[i-1].Hash()
-			// Chain broke ancestry, log a message (programming error) and skip insertion
-			log.Error("Non contiguous header insert", "number", chain[i].Number, "hash", hash,
-				"parent", chain[i].ParentHash, "prevnumber", chain[i-1].Number, "prevhash", parentHash)
-
-			return 0, fmt.Errorf("non contiguous insert: item %d is #%d [%x..], item %d is #%d [%x..] (parent [%x..])", i-1, chain[i-1].Number,
-				parentHash.Bytes()[:4], i, chain[i].Number, hash.Bytes()[:4], chain[i].ParentHash[:4])
-		}
-		// If the header is a banned one, straight out abort
-		if BadHashes[chain[i].ParentHash[types.QuaiNetworkContext]] {
-			return i - 1, ErrBannedHash
-		}
-		// If it's the last header in the cunk, we need to check it too
-		if i == len(chain)-1 && BadHashes[chain[i].Hash()] {
-			return i, ErrBannedHash
+
+	return nil
+}
+
+// Trim
+func (hc *HeaderChain) trim(commonHeader *types.Header, startHeader *types.Header) error {
+	parent := startHeader
+	// Delete each header until common is found
+	for {
+		if parent.Hash() == commonHeader.Hash() {
+			break
 		}
-	}
 
-	// Generate the list of seal verification requests, and start the parallel verifier
-	seals := make([]bool, len(chain))
-	if checkFreq != 0 {
-		// In case of checkFreq == 0 all seals are left false.
-		for i := 0; i <= len(seals)/checkFreq; i++ {
-			index := i*checkFreq + hc.rand.Intn(checkFreq)
-			if index >= len(seals) {
-				index = len(seals) - 1
-			}
-			seals[index] = true
+		// Delete the header and the block
+		rawdb.DeleteHeader(hc.headerDb, parent.Hash(), parent.Number64())
+		hc.bc.Trim(parent)
+
+		parent = hc.GetHeader(parent.Parent(), parent.Number64()-1)
+
+		if parent == nil {
+			log.Warn("unable to trim blockchain state, one of trimmed blocks not found")
+			return nil
 		}
-		// Last should always be verified to avoid junk.
-		seals[len(seals)-1] = true
 	}
+	return nil
+}
 
-	abort, results := hc.engine.VerifyHeaders(hc, chain, seals)
-	defer close(abort)
+// findCommonHeader
+func (hc *HeaderChain) findCommonHeader(header *types.Header) *types.Header {
 
-	// Iterate over the headers and ensure they all check out
-	for i := range chain {
-		// If the chain is terminating, stop processing blocks
-		if hc.procInterrupt() {
-			log.Debug("Premature abort during headers verification")
-			return 0, errors.New("aborted")
-		}
-		// Otherwise wait for headers checks and ensure they pass
-		if err := <-results; err != nil {
-			return i, err
+	for {
+		canonicalHash := rawdb.ReadCanonicalHash(hc.headerDb, header.Number64())
```

Comment:
I think we need a mutex here. Canonical state can change under our feet while we iterate the chain